libc/src/setjmp/x86_64/longjmp.cpp

```diff
@@ -7,15 +7,34 @@
 //===----------------------------------------------------------------------===//
 
 #include "src/setjmp/longjmp.h"
+#include "include/llvm-libc-macros/offsetof-macro.h"
 #include "src/__support/common.h"
 #include "src/__support/macros/config.h"
 
-#if !defined(LIBC_TARGET_ARCH_IS_X86_64)
+#if !defined(LIBC_TARGET_ARCH_IS_X86)
 #error "Invalid file include"
 #endif
 
 namespace LIBC_NAMESPACE_DECL {
 
+#ifdef __i386__
+[[noreturn]]
+LLVM_LIBC_FUNCTION(void, longjmp, (jmp_buf buf, int val)) {
+  asm(R"(
+      mov %[ebx], %%ebx
+      mov %[esi], %%esi
+      mov %[edi], %%edi
+      mov %[ebp], %%ebp
+      mov %[esp], %%esp
+
+      jmp *%[eip]
+      )" ::[ebx] "m"(buf->ebx),
+      [esi] "m"(buf->esi), [edi] "m"(buf->edi), [ebp] "m"(buf->ebp),
+      [esp] "m"(buf->esp), [eip] "m"(buf->eip), [val] "a"(val == 0 ? 1 : val));
+  __builtin_unreachable();
+}
+#else
+[[noreturn]]
 LLVM_LIBC_FUNCTION(void, longjmp, (jmp_buf buf, int val)) {
```

Comment:
(Of course this review isn't about the x86-64 longjmp, but just I see it's also broken similarly.)

Reply:
Yeah, the generated asm is pretty bad.  There's an indirect jump (i.e. `noreturn`), then pops restoring clobbered registers, then a return. i.e. a bit of dead code.

Reply:
https://github.com/llvm/llvm-project/pull/112581